app/src/main/java/org/mozilla/fenix/FenixApplication.kt

```diff
@@ -75,20 +76,21 @@ open class FenixApplication : Application() {
 
     @CallSuper
     open fun setupInMainProcessOnly() {
-        setupMegazord()
-
-        // We want rust logging to go through the log sinks.
-        // This has to happen after initializing the megazord.
-        RustLog.enable()
-
+        var megazordSetup = setupMegazord()
         setDayNightTheme()
-
         registerRxExceptionHandling()
         enableStrictMode()
 
         // Make sure the engine is initialized and ready to use.
         components.core.engine.warmUp()
 
+        // Just to make sure it is impossible for any application-services pieces
+        // to invoke parts of itself that require complete megazord initialization
+        // before that process completes, we wait here, if necessary.
+        if (!megazordSetup.isCompleted) {
+            runBlocking { megazordSetup.await(); }
+        }
```

Comment:
So if we ever add any code, that may call into Rust AS code, between `setupMegazord()` and the `await` here, then this could potentially cause issues (if the setup has not been completed by that time), is that right? Unfortunately that is hard to even know and the compiler can't protect us against this mistake. But I assume we are seeing a performance gain that is worth the risk?

Reply:
> So if we ever add any code, that may call into Rust AS code, between `setupMegazord()` and the `await` here, then this could potentially cause issues (if the setup has not been completed by that time), is that right? Unfortunately that is hard to even know and the compiler can't protect us against this mistake. 

That is absolutely correct. I think that the best we can do is put in some very, very explicit comments. See below.

> But I assume we are seeing a performance gain that is worth the risk?

We do see a performance gain, but *I* am not the person who can make the decision about risk/reward ratio! :-)

Reply:
It shouldn't be that bad. If the call has to do networking, that will fail until the networking part of this runs, and it won't perform any logging (well, megazord setup itself will probably log, but after that, not a peep), until the logging part. But in general failing to set up these things shouldn't be that bad.

Reply:
> We do see a performance gain, but I am not the person who can make the decision about risk/reward ratio! 

I am just curious, but can you share what the performance gains are we see here or why `RustHttpConfig.setClient` and `RustLog.enable()` are "expensive" and worth moving to a different thread?

Reply:
To be clear, all for it if it makes startup faster, but would be good to have an idea how much :).

Reply:
Of the 193ms it takes to execute setupInMainProcessOnly under ND, setClient() takes 22ms. 

For a single iteration of this on ND, the cumulative gains are

`setupInMainProcessOnly`: 193ms vs 174ms
`FenixApplication.onCreate`: 203ms vs 190ms

We do not regain the entire 22ms because it takes a certain amount of time to create the IO dispatchers for the first time. I am thinking about how we could potentially hide that cost as well. 

I will run additional tests on the Pixel hardware I have at my desk and update this accordingly.

Reply:
I did 100 runs of Fenix before/after the change here. In the course of those invocations on Fenix, there are anomalies (startup taking "negative" time) and I threw out those for both the before and after. The result was 71 and 75 "valid" Fenix startups. I averaged those startup times and calculated the percent difference. The result: 4.5%.